src/simulations/collect.jl

```diff
@@ -0,0 +1,361 @@
+###################################################
+# Definition of the data collection API
+###################################################
+get_data(a, s::Symbol) = getproperty(a, s)
+get_data(a, f::Function) = f(a)
+
+# TODO: `add_data` is a misleading name, we need a better one.
+add_data(s, model, when::AbstractVector) = s ∈ when
+add_data(s, model, when::Bool) = when
+add_data(s, model, when) = when(model, s)
+
+"""
+    run!(model, agent_step! [, model_step!], n; kwargs...) → agent_df, model_df
+
+Run the model (step it with the input arguments propagated into `step!`) and collect
+data specified by the keywords, explained one by one below. Return the data as
+two `DataFrame`s, one for agent-level data and one for model-level data.
+
+## Data-deciding keywords
+`agent_properties::Vector` decides the agent data. If an entry is a `name::Symbol`,
+then the data for this entry is agent's field with that `name`. If an entry is a `f::Function`,
+then the data for this entry is just `f(a)` for each agent `a`.
+
+`model_properties::Vector` works exactly like `agent_properties` but for model level data.
+
+`aggregation_dict` decides whether the agent data should be aggregated over agents.
+
+#TODO: Describe what `aggregation_dict` is.
+
+# TODO: Well, how do you use aggregation on entries of `agent_properties` that are
```

Comment:
I am not sure what you mean here. Aggregation happens on dataframe columns.

Reply:
Where? Where are you referencing.I don't know if you mean the first sentence or the second. You can reference multiple lines in the Github code commentiong functionality (just select many of them)

Reply:
I was talking about the second point. The `aggregate_data` function accepts only a dataframe and a dictionary. So, I don't see what you mean by agent_properties that are function.

Reply:
A user defines `f(a) = agent.something ^ 2`. And collects data with `agent_properties = [f]`. What is `aggregate_data` in this case if you want to aggregate over the mean of the function `f`?

Reply:
In this case, it will be `aggregate(df, Dict(Symbol("f(something)")=> mean))`.
Aggregation occurs over the column name that holds values of `f`.

Reply:
perfect.